### PR TITLE
Idempotent Consumer Microservice Pattern #2683

### LINKED ISSUE #2683: Microservice pattern: Idempotent consumer
**Description:**
The Idempotent Customer design pattern is essential in microservices architectures to ensure that repeated processing of the same request does not produce different outcomes. This pattern helps in avoiding duplicate operations in distributed systems where network issues or retries might cause the same request to be sent multiple times.

**Main Elements of the Pattern:**
1. **Unique Request Identifier:** Assign a unique identifier to each request to track and manage its state.
2. **Request Storage:** Store the request data and its state to check for duplicates.
3. **State Management:** Implement mechanisms to manage the state of requests (e.g., pending, processing, completed) and to handle retries gracefully.
4. **Response Handling:** Ensure that the same response is returned for repeated requests, preventing inconsistent outcomes.
5. **Concurrency Control:** Use locking mechanisms or transactional processes to manage concurrent requests and ensure data integrity.

**References:**
- [Microservices Architecture - Idempotent Operations](https://martinfowler.com/bliki/IdempotentOperations.html)
- [Microservices Patterns - Chris Richardson](https://microservices.io/patterns/fault-tolerance/idempotent-consumer.html)
- [Project Contribution Guidelines](https://github.com/iluwatar/java-design-patterns/wiki)

**Acceptance Criteria:**
1. Implement a unique request identifier mechanism for the Idempotent Customer design pattern.
2. Develop a request storage and state management system to handle and track the state of requests.
3. Ensure that repeated requests return consistent responses and implement concurrency control mechanisms to maintain data integrity.

### PR DESCRIPTION
**Pull request by dmgodoy**
## What problem does this PR solve?
Adds idempotent consumer microservice pattern 

Close #2683

### CODE REVIEW

File: idempotent-consumer/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+> In a payment processing system, ensuring that payment messages are idempotent prevents duplicate transactions. For example, if a user’s payment message is accidentally processed twice, the system should recognize the second message as a duplicate and prevent it from executing a second time. By storing unique identifiers for each processed message, such as a transaction ID, the system can skip any duplicate messages. This ensures that a user is not charged twice for the same transaction, maintaining system integrity and customer satisfaction.
+
+### In Plain Words
+
+> The Idempotent Consumer pattern prevents duplicate messages from causing unintended side effects by ensuring that processing the same message multiple times results in the same outcome. This makes message processing safe in distributed systems where duplicates may occur.
+
+### Wikipedia says
+
+> In computing, idempotence is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.
+
```
Comment (iluwatar) [inline idempotent-consumer/README.md:35]: We should add programmatic example section that shows minimal code snippets to implement the pattern

File: idempotent-consumer/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+---
```
Comment (iluwatar) [inline idempotent-consumer/README.md:1]: To be in line with other microservice patterns, we should change the top level folder name to `microservices-idempotent-consumer`

File: idempotent-consumer/pom.xml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                <version>3.2.3</version>
+                <scope>import</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.hibernate</groupId>
+                <artifactId>hibernate-core</artifactId>
+                <version>6.4.4.Final</version>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
```
Comment (iluwatar) [inline idempotent-consumer/pom.xml:54]: Not sure if we need `<dependencyManagement>` section in this Maven submodule. I think it should be used only on parent level.

In the submodule’s pom.xml, you then simply declare the dependency (without specifying the version), and Maven will resolve it using the version defined in the parent’s `<dependencyManagement>`.

I hope it makes sense?

**Review by iluwatar — CHANGES_REQUESTED** [review]
Looks really good! Just a couple of things and then we can proceed with the merge.

**Comment by Ww2Zero** [comment]
您好！您的邮件我已收到。

File: idempotent-consumer/README.md (hunk 1 above)
Comment (dmgodoy) [reply idempotent-consumer/README.md]: Just added a programmatic example for the pattern

File: idempotent-consumer/README.md (hunk 2 above)
Comment (dmgodoy) [reply idempotent-consumer/README.md]: done!

File: idempotent-consumer/pom.xml (hunk 3 above)
Comment (dmgodoy) [reply idempotent-consumer/pom.xml]: I tried this but unfortunately I was not able to make Spring Boot 2.7.5 with Slf4j 2.0.16 and Logback 1.5.6.

I get NoClassDefFoundError: org/slf4j/impl/StaticLoggerBinder in SpringApplication.run method.

I have noticed some other projects using dependencyManagement in the modules:

- health-check
- single-table-inheritance
- layered-architecture
- microservice-aggregator
- microservice-distributed-tracing
- page-controller

Maybe someone has some input on this?

Comment (iluwatar) [reply idempotent-consumer/pom.xml]: Ok, it's not a showstopper. Related issue https://github.com/iluwatar/java-design-patterns/issues/2407. I think we need another refactoring ticket to upgrade the parent level Spring Boot version.

**Review by iluwatar — APPROVED** [review]

**Comment by iluwatar** [comment]
Looks good! Thank you for the contribution 🎉

@all-contributors please add @dmgodoy for code

**Comment by dmgodoy** [comment]
Great! Thanks!

**Pull request merged**